PostgreSQL data connector PR

spsocprime opened the pull request:
Change(s):
   - Added PostgreSQL data connector
   - Added Parsers
   - Added Sample Data

spsocprime commented:
Error in `DataConnectorValidations` check is due to unmatched template but this connector is not of REST_API type
<img width="534" alt="image" src="https://user-images.githubusercontent.com/94110440/165739913-63d5d509-4830-42fb-947d-2425775fc79c.png">

oshezaf reviewed (changes requested):
Comments about the ASIM parser:
- A filtering parser is required.
- Testing with the ASIM testers is required:
  - I ran the tests and found the following issues:
    - Missing EventScheman field.
    - Unsupported value for EventProduct and EventVendor. Obviously, one is needed, but you need to pre-register it with me.
    - Unsupported value for EventType. Should be "Logon".
    - Missing recommended fields that per my analysis should be available: EventResultDetails (based on the type of failure the event denotes), DvcIpAdde or DvcHostname (Computer field)
  - The test data includes only one of the events parsed. At least one sample of each parsed event is needed for testing.

spsocprime commented:
> 

@oshezaf could you please suggest/pre-register `EventProduct` and `EventVendor` values for this parser to be included in our next commit along with the other fixes while addressing your comments?

oshezaf commented:
I have added PostgreSQL as both EventVendor and EventProduct to the test suite. There seems not to be a better option. Documenation update will happen in a few days.

oshezaf commented:
I have added PostgreSQL as both EventVendor and EventProduct to the test suite. There seems not to be a better option. The one product that

spsocprime commented:
> Comments about the ASIM parser:
> 
> * A filtering parser is required.
> * Testing with the ASIM testers is required:
>   
>   * I ran the tests and found the following issues:
>     
>     * Missing EventScheman field.
>     * Unsupported value for EventProduct and EventVendor. Obviously, one is needed, but you need to pre-register it with me.
>     * Unsupported value for EventType. Should be "Logon".
>     * Missing recommended fields that per my analysis should be available: EventResultDetails (based on the type of failure the event denotes), DvcIpAdde or DvcHostname (Computer field)
>   * The test data includes only one of the events parsed. At least one sample of each parsed event is needed for testing.

* Added filtering parser
* Added `EventSchema` field (it was missing as currently it is not in the Mandatory fields [according to the documentation it will become Mandatory on July 1st 2022]
* `EventProduct` and `EventVendor` values were pre-registered
* Fixed `EventType` value
* Added recommended fields `EventResultDetails`, `DvcIpAddr`, `DvcHostname`
* Updated sample data file

oshezaf commented on Parsers/ASimAuthentication/ProductParsers/ASimAuthenticationPostgreSQL.yaml:
```diff
@@ -27,14 +27,17 @@ ParserQuery: |
     , TargetUsernameType = 'Simple'
     , TargetUsername = extract(@'user=(.*?)\sdatabase', 1, RawData)
-    , EventMessage = 'Connection authorized'
+    , EventResultDetails = 'Connection authorized'
```
EventResultDetails has to be mapped to one of the [allowed values](https://docs.microsoft.com/azure/sentinel/authentication-normalization-schema#common-fields-with-specific-guidelines)

oshezaf commented on Parsers/ASimAuthentication/ProductParsers/ASimAuthenticationPostgreSQL.yaml:
```diff
@@ -108,18 +120,22 @@ ParserQuery: |
-  union isfuzzy=true PostgreSQLSignInAuthorized(disabled), PostgreSQLAuthFailure1(disabled), PostgreSQLAuthFailure2(disabled), PostgreSQLAuthFailure3(disabled), PostgreSQLDisconnect(disabled)
\ No newline at end of file
+  union isfuzzy=true PostgreSQLSignInAuthorized(disabled), PostgreSQLAuthFailure1(disabled), PostgreSQLAuthFailure2(disabled), PostgreSQLAuthFailure3(disabled), PostgreSQLDisconnect(disabled)
+  | project-away Computer, MG, ManagementGroupName, RawData, SourceSystem, TenantId
```
**project-away** in each branch of the union rather than after the union. It has dramatic effect on memory use.

spsocprime replied:
fixed

oshezaf commented on Parsers/ASimAuthentication/ProductParsers/ASimAuthenticationPostgreSQL.yaml:
```diff
@@ -27,14 +27,17 @@ ParserQuery: |
     , TargetUsernameType = 'Simple'
     , TargetUsername = extract(@'user=(.*?)\sdatabase', 1, RawData)
-    , EventMessage = 'Connection authorized'
+    , EventResultDetails = 'Connection authorized'
```
**EventResultDetails** has to be mapped to one of the [allowed values](https://docs.microsoft.com/azure/sentinel/authentication-normalization-schema#common-fields-with-specific-guidelines). The value your generated should go to **EventOriginalRestultDetails**

spsocprime replied:
There are allowed values for `EventResultDetails` only for failed events. Where can we find allowed values for `EventResultDetails` for successful logon events?

oshezaf replied:
There aren't any define and per the current definition it should be empty for Success. (so use only EventOriginalResultDetails). Thar said, if you feel I need to define values, I can. What would they be?

spsocprime replied:
fixed

oshezaf commented on Parsers/ASimAuthentication/ProductParsers/vimAuthenticationPostgreSQL copy.yaml:
```diff
@@ -0,0 +1,205 @@
+Parser:
```
Remove **copy** from the file name.

spsocprime replied:
fixed

oshezaf commented on Parsers/ASimAuthentication/ProductParsers/vimAuthenticationPostgreSQL copy.yaml:
```diff
@@ -0,0 +1,205 @@
+    , TargetUsername = extract(@'user=(.*?)\sdatabase', 1, RawData)
+    , EventResultDetails = 'Connection authorized'
+  // ********************** <Postfilterring> **********************************
+    | where (targetusername_has=='*' or TargetUsername has targetusername_has)
```
As part of the prefiltering you need to check if the username appears in the RawData field. The postfiltering than makes sure that this was indeed the username.

spsocprime replied:
fixed

oshezaf commented on Parsers/ASimAuthentication/ProductParsers/vimAuthenticationPostgreSQL copy.yaml:
```diff
@@ -0,0 +1,205 @@
+Description: |
+      This Query Parser maps PostgreSQL sign in logs to the Advanced Security Information Model authenticaion schema.
+ParserName: ASimAuthenticationPostgreSQL
+ParserParams:
```
The YAML file needs to include the additional parameters as well.

spsocprime replied:
fixed

oshezaf commented on Parsers/ASimAuthentication/ProductParsers/vimAuthenticationPostgreSQL copy.yaml:
```diff
@@ -0,0 +1,205 @@
+                  ,  PostgreSQLAuthFailure2(starttime, endtime, targetusername_has, disabled)
+                  ,  PostgreSQLAuthFailure3(starttime, endtime, targetusername_has, disabled)
+                  ,  PostgreSQLDisconnect(starttime, endtime, targetusername_has, disabled)
+  | project-away Computer, MG, ManagementGroupName, RawData, SourceSystem, TenantId
```
See ASIM parser comment.

spsocprime replied:
fixed

oshezaf commented on Parsers/ASimAuthentication/ProductParsers/vimAuthenticationPostgreSQL copy.yaml:
```diff
@@ -0,0 +1,205 @@
+    , DvcIpAddr = extract(@'\d{1,3}\.\d{1.3}\.\d{1,3}\.\d{1,3}', 1, Computer)
+    , TargetUsernameType = 'Simple'
+    , TargetUsername = extract(@'user=(.*?)\sdatabase', 1, RawData)
+    , EventResultDetails = 'Connection authorized'
```
See ASIM parser comment

oshezaf commented on Parsers/ASimAuthentication/ProductParsers/ASimAuthenticationPostgreSQL.yaml:
```diff
@@ -0,0 +1,149 @@
+    , TargetUsernameType = 'Simple'
+    , TargetUsername = extract(@'user=(.*?)\sdatabase', 1, RawData)
+    , SrcIpAddr = extract(@'host=\[?(.*?)\]?', 1, RawData)
+    , EventResultDetails = 'Session closed'
```
Should be 'Session expired'

spsocprime replied:
fixed

oshezaf reviewed (changes requested):
Two additional comments. One of them is created by the changes. It does point that you need to try running and test again against that asim testers after each change round.

spsocprime commented on Parsers/ASimAuthentication/ProductParsers/ASimAuthenticationPostgreSQL.yaml:
```diff
@@ -0,0 +1,149 @@
+  // **** Aliases
+  | extend
+          User=TargetUsername
+        , Dvc=Computer
```
fixed

oshezaf reviewed (changes requested):
It works thanks. One last, even though it does not change this parser, it is an important best practice: please use isfuzzy=false in all parsers (ASIM or not)

isfuzzy=true is needed if you expect an error in one of the unioned sub-queries, typically when unioning over different tables which only some may be available in the workspace.

In the current use case, it just hides errors when testing.

mkchiliveri commented on Solutions/PostgreSQL/Data Connectors/Connector_PostgreSQL.json:
```diff
@@ -0,0 +1,168 @@
+    "title": "PostgreSQL Events",
+    "publisher": "PostgreSQL",
+    "descriptionMarkdown": "PostgreSQL data connector provides the capability to ingest [PostgreSQL](https://www.postgresql.org/) events into Microsoft Sentinel. Refer to [PostgreSQL documentation](https://www.postgresql.org/docs/current/index.html) for more information.",
+    "additionalRequirementBanner": "This data connector depends on a parser based on Kusto Function to work as expected. Follow the steps to use this Kusto Function alias **PostgreSQLEvent** in queries and workbooks. [Follow steps to get this Kusto Function>](https://aka.ms/sentinel-postgresql-parser)",
```
Please validate and provide proper link as it is redirecting to bing https://aka.ms/sentinel-postgresql-parser

spsocprime replied:
Such links are handled by the Microsoft team. We just put the proper text into data connector template

v-sabiraj replied:
@spsocprime, @mkchiliveri, handled and its working, can you please check, thanks.

mkchiliveri commented on Solutions/PostgreSQL/Data Connectors/Connector_PostgreSQL.json:
```diff
@@ -0,0 +1,168 @@
+        "description":"It may take upwards of 20 minutes until your logs start to appear in Microsoft Sentinel."
+    }
+]
+}
```
Please add the metadata for data connector

spsocprime replied:
metadata added

mkchiliveri reviewed (approved):
Looks good, Approved

anki-narravula reviewed (approved):
I see it is approved by Mahesh and Ofer already.
